### PR TITLE
[spec] Support saved queries for selectURL

### PR DESCRIPTION
**Pull request by pythagoraskitty**
We update the spec to support saved queries, as a followup to #176.

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/188.html" title="Last updated on Sep 20, 2024, 5:14 PM UTC (fa8183f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/shared-storage/188/0ebd7a5...fa8183f.html" title="Last updated on Sep 20, 2024, 5:14 PM UTC (fa8183f)">Diff</a>

### CODE REVIEW

**Comment by pythagoraskitty** [comment]
@xyaoinum Could you please take a first pass? I won't land until after #176 lands and after I have a spec mentor review also. Thanks!

File: spec.bs
Comment (xyaoinum) [inline spec.bs:1288]: I think usually we omit the `= ""` for string default values.

**Review by xyaoinum — APPROVED** [review]
LGTM % a nit

File: spec.bs
Comment (pythagoraskitty) [reply spec.bs]: Done.

**Comment by pythagoraskitty** [comment]
@wanderview Please take a look, thanks!

File: spec.bs
Comment (wanderview) [inline spec.bs:336]: Is "nonempty" a spec term?  Is there somewhere else that is using this?  Maybe you want to say `is a string that is not the empty string`?

File: spec.bs
Comment (wanderview) [inline spec.bs:338]: It appears that you could use the same name for two completely different queries and this would return the value of the first query.  Is this the desired behavior?

For example, maybe we should return an error if the parameters passed to selectUrl in the second call differ from the first call?  Or should the second call override the parameters from the first selectUrl by creating a new saved query?

This is kind of stemming from the awkardness of adding a parameter to an existing function call.  It seems like what you might want is something more like how SQL has separate "prepare statement" and "execute statement" calls.  You could consider a `prepareSelectUrl()` that takes all params and assigns a name.  Then a separate `selectUrl(name)` call that operates only on the name without any other params.

If we want to keep a single function call that is overloaded I think we should just be intentional if the site calls it twice with different arguments, but the same name.

**Review by wanderview — CHANGES_REQUESTED** [review]

File: spec.bs
Comment (pythagoraskitty) [reply spec.bs:338]: > It appears that you could use the same name for two completely different queries and this would return the value of the first query. Is this the desired behavior?

The short answer is: probably yes.  

The long answer is that I did forget to take into account the operation name, and I may not want to ignore that. I am thinking about that some more offline.  

But we deliberately want the array of URLs to be able to be different, and note that we do handle the error that happens if we end up with an out of range index because of this.

We would ignore any passed `data` for `options` in the case that we are able to retrieve a saved query result. The other `options` parameters are still handled.

So there shouldn't be a need for a separate `prepareSelectUrl()` statement.

Comment (wanderview) [reply spec.bs:338]: In that case I recommend returning an error if the operation name doesn't match the previously used operation name for the same saved query.  Its probably a developer mistake.  Allowing the same saved query name for different operations sounds like a footgun.

File: spec.bs
Comment (pythagoraskitty) [reply spec.bs]: Done.

File: spec.bs
Comment (pythagoraskitty) [reply spec.bs]: After further contemplation and discussion with the team, we've decided two things.

1. Saved queries will be stored on a per-origin (instead of per-site) basis. I have updated this spec PR accordingly.
2. We do indeed want names of saved queries to be independent of operation names, and for the storing and lookup of saved queries to ignore the operation names. So two calls to `selectURL()` that have different operation names could still access the same saved query result if they use the same `savedQuery` parameter. While this could lead to some developers making mistakes, it also allows for greater flexibility and simplicity. We would need to be sure to document this properly, of course, including in the current explainer PR (where the effect of differing operation names is currently not mentioned).

Comment (jkarlin) [reply spec.bs]: I don't have particularly strong feelings here. It seems unlikely to be a footgun to me, but at the same time I can't think of a good reason to use the same saved-query on two different operations. So, let's say that we go ahead with your proposal here as it seems reasonable.

So now the proposal is that a saved-query key is `<origin, operation_name, saved_query_name>`. 

For similar reasons, what would you think about also including the worklet script url as part of the key?

e.g., `<origin, operation_name, worklet_script_url, saved_query_name>`

Like operations, presumably different scripts could have different behavior, even if the operation name is the same. This gets a little funky in that urls aren't actually const resources, but still.

Comment (pythagoraskitty) [reply spec.bs]: OK, I have updated the spec so that queries are keyed by `<origin, operation_name, saved_query_name>`.

I think we should just not reuse the query if there is different operation name, rather than throwing, to simplify things.

Comment (wanderview) [reply spec.bs]: Just to be clear, keying on `<origin, operation_name, saved_query_name>` does your proposal above, right?  It ignores mismatched operations because you end up with different slots in your map.  To do what I was proposing I think the key needs to be `<origin, saved_query_name>` and then in the map value the `operation_name` is saved so you can compare it on lookup.

I'm not sure I understand enough about how different worklets can mix on a page to have an opinion on how to treat conflicting scripts URLs.  If we assume folks are going to hash URLs for different versions of the same worklet, though, it does seem like a flow we should be concerned with.

Comment (pythagoraskitty) [reply spec.bs]: > Just to be clear, keying on <origin, operation_name, saved_query_name> does your proposal above, right? It ignores mismatched operations because you end up with different slots in your map. To do what I was proposing I think the key needs to be <origin, saved_query_name> and then in the map value the operation_name is saved so you can compare it on lookup.

Yes. The key would need to be `<origin, saved_query_name` to do what you propose. But I think it would be best to key on `<origin, operation_name, saved_query_name>` instead because it seems to make more sense for us to simply narrow the scope for the saved queries further from the beginning, then to add additional errors. We wouldn't throw if there is a mismatch, but we wouldn't re-use a query in the case of a mismatch either.

File: spec.bs
Comment (wanderview) [inline spec.bs:1008]: Rather than a map containing a map that contains a further map, you can instead use a tuple as the map key.  For example:

```
[=map/Set=] |topLevelTraversable|'s [=page budget=]'s [=saved query map=][(|origin|, |operationName|, |savedQueryName|)] to |index|.
```

**Review by wanderview — APPROVED** [review]
LGTM with one spec language improvement fixed.  This changes the setting and getting of the saved query using tuples as the key.

File: spec.bs
Comment (wanderview) [reply spec.bs:1008]: An example of tuples as keys here: https://html.spec.whatwg.org/multipage/webappapis.html#module-map

Comment (pythagoraskitty) [reply spec.bs]: Done.

Comment (pythagoraskitty) [reply spec.bs]: Done.

Comment (wanderview) [reply spec.bs]: LGTM

**Comment by pythagoraskitty** [comment]
After further discussion with the team, we decided to also key on the script's URL. So now I have updated the spec so the the key is `<origin, script URL, operation name, query name>`.

Would you mind please taking one more look? @wanderview @xyaoinum 

Thanks!

File: spec.bs
Comment (wanderview) [inline spec.bs:958]: I recommend phrasing this like `a [=string=] <dfn for="saved query map">operation name</dfn>, a [=string=] <dfn for="saved query map">saved query name< /dfn>`, etc.  So folks can tell the difference between the strings, etc.  I would do it for the URL and origin as well for completeness.

**Comment by wanderview** [comment]
> After further discussion with the team, we decided to also key on the script's URL. So now I have updated the spec so the the key is `<origin, script URL, operation name, query name>`.
> 
> Would you mind please taking one more look? @wanderview @xyaoinum
> 
> Thanks!

I added one more comment, but otherwise LGTM.

**Comment by xyaoinum** [comment]
Still LGTM

File: spec.bs
Comment (pythagoraskitty) [reply spec.bs]: Done.

**Pull request merged**
